polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1251,6 +1367,65 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
+	fn do_reserve_withdraw_assets(
+		origin: OriginFor<T>,
+		dest: Box<VersionedMultiLocation>,
+		beneficiary: Box<VersionedMultiLocation>,
+		assets: Box<VersionedMultiAssets>,
+		fee_asset_item: u32,
+		maybe_weight_limit: Option<WeightLimit>,
+	) -> DispatchResult {
+		let origin_location = T::ExecuteXcmOrigin::ensure_origin(origin)?;
+		let dest = (*dest).try_into().map_err(|()| Error::<T>::BadVersion)?;
+		let beneficiary: MultiLocation =
+			(*beneficiary).try_into().map_err(|()| Error::<T>::BadVersion)?;
+		let assets: MultiAssets = (*assets).try_into().map_err(|()| Error::<T>::BadVersion)?;
+
+		ensure!(assets.len() <= MAX_ASSETS_FOR_TRANSFER, Error::<T>::TooManyAssets);
```

Comment:
We have `T::XcmReserveTransferFilter` for opposite direction,
so maybe we could add here also some filter for (dest, assets) e.g. `T::XcmReserveWithdrawFilter` that runtime could have ability to check/control that particular assets are allowed for initiate reserve withdraw for the requested destination (e.g. `Everything`, `Nothing`, `()`, or some AssetId-MultiLocation-based custom filter...).

The processing of `ReserveAssetDeposited` is controlled by [`xcm_executor::Config::IsReserve`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L488-L496) filter which means that we trust to configured reserve location with particular assets.

So maybe the runtime would like to have ability to set `T::XcmReserveWithdrawFilter` with the same/similar filter like `IsReserve`, which would mean "I trust to XYZ reserve location, so I only allow to initiate reserve withdraw if destination is this XYZ reserve location".

Reply:
Having a reserve-transfer filter makes sense, and failing the filter leads to failed extrinsic but no asset loss - the transfer will just not go through.

For the opposite direction (this PR), when attempting to withdraw some reserve-based asset using this new extrinsic, the _correct usage_ assumption is that local chain has already previously received “wTOK” based on reserved “TOK” on reserve chain. Filtering at this point only leads to “locking/blocking” retrieval of original reserve assets from the reserve account.

And if we really wanted to filter receival of reserve-based assets, it should probably be done at the xcm executor level on “ReserveAssetDeposited” and “InitiateReseveWithdraw” instructions.

Reply:
> For the opposite direction (this PR), when attempting to withdraw some reserve-based asset using this new extrinsic, 

We dont know if asset is "reserve-based" asset (the only possible check which comes to my mind is to check `IsReserve`), so e.g. `T::XcmReserveWithdrawFilter` could prevent initiate reserve withdraw of DOTs from `AssetHubPolkadot`, because AssetHubPolkadot is reserve location for DOTs.